lib/rubocop/cop/style/hash_transform_methods.rb

```diff
@@ -0,0 +1,246 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop looks for uses of `_.each_with_object({}) {...}`,
+      # `_.map {...}.to_h`, and `Hash[_.map {...}]` that are actually just
+      # transforming either the keys or the values of a hash, and tries to use
+      # a simpler & faster call to `transform_keys` or `transform_values`
+      # instead.
+      #
+      # This can produce false positives if we are transforming an enumerable
+      # of key-value-like pairs that isn't actually a hash, e.g.:
+      # `[[k1, v1], [k2, v2], ...]`
+      #
+      # This cop should only be enabled on Ruby version 2.5 or newer
+      # (`transform_values` was added in 2.4, and `transform_keys` in 2.5.)
+      #
+      # @example
+      #   # bad
+      #   {a: 1, b: 2}.each_with_object({}) { |(k, v), h| h[k] = v * v }
+      #   {a: 1, b: 2}.map { |k, v| [k.to_s, v] }
+      #
+      #   # good
+      #   {a: 1, b: 2}.transform_values { |v| v * v }
+      #   {a: 1, b: 2}.transform_keys { |k| k.to_s }
+      class HashTransformMethods < Cop
+        extend TargetRubyVersion
+
+        minimum_target_ruby_version 2.5
+
+        EACH_WITH_OBJECT_PATTERN = NodePattern.new(<<~PATTERN)
```

Comment:
Can you use `def_node_matcher` instead of `CONSTANT = NodePattern.new(<<~PATTERN)`?

Reply:
Sure, if that's the standard style, but OOC is there a particular reason for it? AFAICT it does the ~same thing but with a little more magic that I find confusing.